network/packet_listener_proxy.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package network
+
+import (
+	"context"
+	"errors"
+	"net"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-internal-sdk/transport"
+)
+
+const packetMaxSize = 2048
+
+// Compilation guard against interface implementation
+var _ PacketProxy = (*packetListenerProxyAdapter)(nil)
+var _ PacketRequestSender = (*packetListenerRequestSender)(nil)
+
+type packetListenerProxyAdapter struct {
+	listener transport.PacketListener
+	timeout  time.Duration
+}
+
+type packetListenerRequestSender struct {
+	conn    net.PacketConn
+	timeout time.Duration
+}
+
+// NewPacketProxyFromPacketListener creates a new [PacketProxy] that uses the existing [transport.PacketListener].
+// You can use this function if you already have an implementation of [transport.PacketListener] and would like to
+// inject it into one of the network stacks (for example, network/lwip2transport) as UDP traffic handlers.
+func NewPacketProxyFromPacketListener(pl transport.PacketListener) (PacketProxy, error) {
+	if pl == nil {
+		return nil, errors.New("pl must not be nil")
+	}
+	return &packetListenerProxyAdapter{
+		listener: pl,
+		timeout:  30 * time.Second,
```

Comment:
Not for this PR, but we may want to consider a strategy similar to the server.

We use a default of 5 minutes:
https://github.com/Jigsaw-Code/outline-ss-server/blob/989add59c0adb7d44013298f64f7eabdba45371e/cmd/outline-ss-server/main.go#L49

However, for DNS we use 17 seconds:
https://github.com/Jigsaw-Code/outline-ss-server/blob/989add59c0adb7d44013298f64f7eabdba45371e/service/udp.go#L275

Please read "Mapping Refresh" on RFC 4787: https://datatracker.ietf.org/doc/html/rfc4787#section-4.3